pkgs/by-name/README.md

```diff
@@ -0,0 +1,107 @@
+# Name-based package directories
+
+The structure of this directory maps almost directly to top-level package attributes.
+This is the recommended way to add new top-level packages to Nixpkgs [when possible](#limitations).
+
+## Example
+
+The top-level package `pkgs.some-package` may be declared by setting up this file structure:
+
+```
+pkgs
+└── by-name
+   ├── so
+   ┊  ├── some-package
+      ┊  └── package.nix
+
+```
+
+The `package.nix` may look like this:
+
+```nix
+# A function taking an attribute set as an argument
+{
+  # Get access to top-level attributes for use as dependencies
+  lib,
+  stdenv,
+  libbar,
+
+  # Make this derivation configurable using `.override { enableBar = true }`
+  enableBar ? false,
+}:
+
+# The return value must be a derivation
+stdenv.mkDerivation {
+  # ...
+  buildInputs =
+    lib.optional enableBar libbar;
+}
+```
+
+You can also split up the package definition into more files in the same directory if necessary.
+
+Once defined, the package can be built from the Nixpkgs root directory using:
+```
+nix-build -A some-package
+```
+
+See the [general package conventions](../README.md#conventions) for more information on package definitions.
+
+### Changing implicit attribute defaults
+
+The above expression is called using these arguments by default:
+```nix
+{
+  lib = pkgs.lib;
+  stdenv = pkgs.stdenv;
+  libbar = pkgs.libbar;
+}
+```
+
+But if the package needs e.g. `pkgs.libbar_2` instead, you can change the default `libbar` attribute argument by adding a definition as follows to [`pkgs/top-level/all-packages.nix`](../top-level/all-packages.nix):
+```nix
+libfoo = callPackage ../by-name/so/somePackage/package.nix {
+  libbar = libbar_2;
```

Comment:
Since `libbar_2` would be imported in `by-name-overlay.nix`, it would be

```suggestion
  libbar = pkgs.libbar_2;
```

right?

Reply:
Oh, wait, it's somePackage that's in by-name, not `libbar`. And we're already not using `pkgs.` for versioned packages.

Reply:
There's effectively no difference whether something is defined using `pkgs/by-name` or anywhere else. And also in `all-packages.nix`, there's a `with pkgs;` at the top, so `libbar_2 == pkgs.libbar_2` :)